dmipy/core/modeling_framework.py

```diff
@@ -2191,6 +2193,383 @@ def __call__(self, acquisition_scheme, **kwargs):
         return E
 
 
+class MultiCompartmentAMICOModel(MultiCompartmentModelProperties):
+    def __init__(self, models, S0_tissue_responses=None, parameter_links=None):
+        self.models = models
+        self.N_models = len(models)
+        if S0_tissue_responses is not None:
+            if len(S0_tissue_responses) != self.N_models:
+                msg = 'Number of S0_tissue responses {} must be same as '\
+                      'number of input models {}.'
+                raise ValueError(
+                    msg.format(len(S0_tissue_responses), self.N_models))
+        self.S0_tissue_responses = S0_tissue_responses
+        self.parameter_links = parameter_links
+        if parameter_links is None:
+            self.parameter_links = []
+
+        self._prepare_parameters()
+        self._prepare_partial_volumes()
+        self._prepare_parameter_links()
+        self._prepare_model_properties()
+        self._check_for_double_model_class_instances()
+        self._prepare_parameters_to_optimize()
+        self._check_for_NMR_and_other_models()
+        self.x0_parameters = {}
+        self._forward_model_matrix = None
+        self._parameter_indices = {}
+
+        if not have_numba:
+            msg = "We highly recommend installing numba for faster function "
+            msg += "execution and model fitting."
+            print(msg)
+        if not have_pathos:
+            msg = "We highly recommend installing pathos to take advantage of "
+            msg += "multicore processing."
+            print(msg)
+
+    def _check_for_NMR_and_other_models(self):
+        model_types = [model._model_type for model in self.models]
+        if "NMRModel" in model_types:
+            if len(np.unique(model_types)) > 1:
+                msg = "Cannot combine 1D-NMR and other 3D model types together"
+                msg += " into a MultiCompartmentAMICOModel."
+                raise ValueError(msg)
+
+    def _check_if_model_orientations_are_fixed(self):
+        # TODO: return True if the orientations are fixed, False o/wise.
+        raise NotImplementedError
+
+    def _create_forward_model_matrix(self):
+        # TODO: implement the function that creates the forward model matrix.
+        #  The function is expected to store the matrix in the
+        #  self._forward_model_matrix field and at the same time it must define
+        #  a dictionary whose keys are the parameter names and the values are
+        #  the set of column indices in the matrix corresponding to the
+        #  parameter. This dictionary must be stored in self._parameter_indices.
+        raise NotImplementedError
+
+    @property
+    def forward_model(self):
+        """Return the forward model matrix associated to the AMICO model"""
```

Comment:
one can only generate the forward model *given* the directions of the model at this voxel. How are you including this here?

Reply:
It reads the directions from the model parameters, that should be set manually before calling anything that needs the forward model matrix

Reply:
I get that, but do you want to pre-calculate the forward models for all unique orientations in the fixed orientations??
This can quickly become a huge chunk of data.

I think this function should calculate a forward model for a single set of parameters on-the-fly, this way it can also be used in the optimizer, generating a new forward_model matrix for each incoming voxel, and the same in the fitted model `predict` function.

This means that the function call should be `def forward_model(self, orientation_parameters)`, since the sampling over the other parameters would be set beforehand. What do you think about this?